SG Redesign

mdwhitten opened the pull request:
Redesigned the SG module to be used more flexibly outside of WLAN specific development.

The primary efforts were focused on isolating driver calls to the specific functions that require them. Hence, `ConfigureInstrument` no longer changes the generation or power mode. Instead, these are set in appropriate downstream functions as required.

The other primary change was introducing a new method for simple endless generation with a trigger, requested by @smooresni and @heyItsDale. Outside of WLAN, the bursted generation code is needlessly complicated for most standard tasks. This change also renamed these bursted generation functions to more accurately reflect their function.

mdwhitten commented:
Hey @smooresni ,

Could you review my changes in this pull request? Outside of a few issues I was fixing, these changes are largely based on our conversations about how the generation code should change. Let me know what you think or if you have any other ideas.

smooresni commented on Source/SG/SG.cs:
```diff
@@ -24,17 +24,18 @@ public struct InstrumentConfiguration
             public double CarrierFrequency_Hz;
             public double AverageInputPower_dBm;
```
Average input power on SG is a little confusing. Suggest we change to DutAverageInputPower_dBm

cluton99 replied:
makes sense to me

smooresni commented on Source/SG/SG.cs:
```diff
@@ -24,17 +24,18 @@ public struct InstrumentConfiguration
             public bool ShareLOSGToSA;
-            public void SetDefaults()
+        }
+        public static InstrumentConfiguration GetDefaultInstrumentConfiguration()
```
why not just make this a static method of the struct? returning self?

mdwhitten replied:
Can you post a comment showing me what you're suggesting in code? You can use the following Markdown syntax with three backticks (\```) to insert a code block formatted properly for C#:
> \```c#
> // code goes here
> \```

i.e.: 
```c#
public static class MyClass {
```

smooresni replied:
```c#
public struct InstrumentConfiguration
        {
            public string ReferenceClockSource;
            public double CarrierFrequency_Hz;
            public double AverageInputPower_dBm;
            public double ExternalAttenuation_dBm;
            public bool ShareLOSGToSA;
            public static InstrumentConfiguration GetDefault()
            {
                return new InstrumentConfiguration()
                {
                    ReferenceClockSource = RfsgFrequencyReferenceSource.PxiClock.ToString(),
                    CarrierFrequency_Hz = 1e9,
                    AverageInputPower_dBm = 0,
                    ExternalAttenuation_dBm = 0,
                    ShareLOSGToSA = true,
                };
            }
        }
```

mdwhitten replied:
Oh, that's actually a really great idea. I originally had it as follows:
```c#
public struct SomeStruct {
   // Values
   public void SetDefaults()
}
```
. But based on your feedback a while ago at the .NET user meeting I changed it to the current syntax. The syntax above allowed the `SetDefaults` method to act on the created struct itself, which was nice (but required multiple lines of code in order to initialize.)

Your suggestion is ideal because it encapsulates both solutions. I'll go with that method, but I won't change it for this pull request because this affects all modules. I'll open an issue and tag you in it.

smooresni commented on Source/SG/SG.cs:
```diff
@@ -68,31 +73,27 @@ public struct PAENConfiguration
-            
-            public void SetDefaults()
+        }
+        public static PAENConfiguration GetDefaultPAENConfiguration()
```
another opportunity for a static GetDefault method

mdwhitten replied:
Resolving due to comment above.

smooresni commented on Source/SG/SG.cs:
```diff
@@ -24,17 +24,18 @@ public struct InstrumentConfiguration
+                CarrierFrequency_Hz = 1e9,
+                AverageInputPower_dBm = 0,
+                ExternalAttenuation_dBm = 0,
+                ShareLOSGToSA = true,
```
why default true instead of false?

cluton99 replied:
The typical use case is PA/FEM DUTs where shared LO is the recommended configuration.

smooresni commented on Source/SG/SG.cs:
```diff
@@ -52,13 +53,17 @@ public struct WaveformTimingConfiguration
             public double DutyCycle_Percent;
```
This feels more like a generation timing configuration rather than attributes of the waveform

mdwhitten replied:
I think that how GitHub was showing this it might have appeared like this was a property of `Waveform` but it's actually a part of `WaveformTimingConfiguration`. If you hit the expand button it will show you the entire code:
![image](https://user-images.githubusercontent.com/47825680/61390116-ccb15f00-a87f-11e9-96c1-0518173c118a.png)

Or check it out below. With this context, do you still think it is located in the incorrect place?

```c#
        public struct Waveform
        {
            public string WaveformName;
            public ComplexWaveform<ComplexSingle> WaveformData;
            public double SignalBandwidth_Hz;
            public double PAPR_dB;
            public double BurstLength_s;
            public double SampleRate;
            public int[] BurstStartLocations;
            public int[] BurstStopLocations;
        }
        public struct WaveformTimingConfiguration
        {
            public double DutyCycle_Percent;
```

smooresni replied:
The way I look at it is "Does the waveform have a duty cycle?" or "Does the test require generation of the waveform conform to a specific duty cycle?" 
I guess the difference is, in WLANG SFP I could set the deadtime to make the waveform have a 50% duty cycle vs during generation i can dynamically configure the instrument to generate the waveform 50% of the time

smooresni commented on Source/SG/SG.cs:
```diff
@@ -68,31 +73,27 @@ public struct PAENConfiguration
-            rfsgHandle.DeviceEvents.MarkerEvents[0].ExportedOutputTerminal =
-                RfsgMarkerEventExportedOutputTerminal.FromString(instrConfig.BurstStartTriggerExportTerminal);
 
             if (instrConfig.ShareLOSGToSA)
```
where is the else case?! what happens if i want to disable LO sharing after I configured it to share the LOs once before

mdwhitten replied:
Yikes, good point

mdwhitten replied:
Fixed in ad04c39ac88e293b6dad2cf48805165c53808402

smooresni commented on Source/SG/SG.cs:
```diff
@@ -147,9 +148,10 @@ public static Waveform LoadWaveformFromTDMS(NIRfsg rfsgHandle, string filePath,
         public static void DownloadWaveform(NIRfsg rfsgHandle, ref Waveform waveform)
```
comments for ReadWaveform:
why are we constructing a waveform when it is going to be replaced anyway? waste of operations
syntactic sugar might be able to reduce number of operations for burst null checks.
suggest we read samplerate from file rather than inverting from waveform timing, rounding errors can occur with current method
burst length calculation hardcoded to only first burst of waveform

smooresni replied:
comments for DownloadWaveform:
having the ref keyword in the function signature might scare a dev that you are going to modify his/her waveform without their consent

mdwhitten replied:
>1. why are we constructing a waveform when it is going to be replaced anyway? waste of operations

You're right. Some reason I thought I had to initialize it but I don't.
>2. syntactic sugar might be able to reduce number of operations for burst null checks.

Agreed; let me know if you can think of something. Doesn't appear to be any oneliner for "is null or empty" for arrays like strings in C#. Checking the array length throws an exception if null (which is what is returned from RFSGPlayback if it doesn't exist).
>3. suggest we read samplerate from file rather than inverting from waveform timing, rounding errors can occur with current method

Interesting. Didn't know that, but sounds fine to me.
>4. burst length calculation hardcoded to only first burst of waveform

Dang; I fixed this in LabVIEW but not here.
>5. having the ref keyword in the function signature might scare a dev that you are going to modify his/her waveform without their consent - this applies to the other function signatures with the ref keyword as well, suggest we remove it

My thought was the waveform could be a hefty set of data to pass around by value, so thought by reference might be less overall. Thoughts? Is this tipping the scale for "optimal" vs. "simple"?

smooresni replied:
the struct is holding a waveform object pointer, so the waveform never gets passed via the stack, only a pointer to the object in the heap gets passed

mdwhitten replied:
Good point on 5. Agreed and removed.

Addressed all of these in commit 476d3426dab68dc7cdd04db6ae5df47725688388. Please review and close conversation if fixed.

smooresni commented on Source/SG/SG.cs:
```diff
@@ -172,10 +174,31 @@ public static void DownloadWaveform(NIRfsg rfsgHandle, ref Waveform waveform)
             NIRfsgPlayback.StoreWaveformRuntimeScaling(rfsgPtr, waveform.WaveformName, -1.5);
```
should this be moved to a waveform property but defaulted to -1.5 in the readfromfile? what is the file has a different runtime scaling?

smooresni commented on Source/SG/SG.cs:
```diff
@@ -172,10 +174,31 @@ public static void DownloadWaveform(NIRfsg rfsgHandle, ref Waveform waveform)
             NIRfsgPlayback.StoreWaveformRuntimeScaling(rfsgPtr, waveform.WaveformName, -1.5);
             NIRfsgPlayback.StoreWaveformRFBlankingEnabled(rfsgPtr, waveform.WaveformName, false);
```
this seems counter-productive. why set the burst start and stop locations in RFSG playback if you are going to hardcode blanking to false?

mdwhitten replied:
Even though blanking is disabled, we still use the database features of RFSGPlayback in the test sequences. For example, when someone is sweeping waveforms in TestStand, I need to retrieve the waveform's properties from the database dynamically. The burst start/stop locations are leveraged for calculating the total length of the waveform; hence, they need to be stored in the database either way for later retrieval.

smooresni replied:
Okay fair enough

smooresni commented:
Some general comments:
- How do we abort continuous generation with the RDL?
- Normalize waveform method needs to be adjusted for waveforms with multiple bursts
- Lets talk about the weird script thing in TogglePFILine and how we can add comments
- GetWaveformParametersByName will beed to be updated for burst length

mdwhitten commented:
/AzurePipelines Run

mdwhitten commented:
/AzurePipelines Run

mdwhitten commented:
@smooresni 
> Some general comments:
> 
> * How do we abort continuous generation with the RDL?

I implemented something like what we discussed for your review in  3b420e2fb86f8f2432563426be978f5d208578df

> * Normalize waveform method needs to be adjusted for waveforms with multiple bursts

Agreed; I think this is out of scope for this particular commit but I will open an issue for this.

> * Lets talk about the weird script thing in TogglePFILine and how we can add comments

Agreed; I think this is out of scope for this particular commit but I will open an issue for this.

> * GetWaveformParametersByName will beed to be updated for burst length

Fixed in 476d3426dab68dc7cdd04db6ae5df47725688388.
